### ISSUE TITLE
CS8619: Missmatchning target type when having nested function with NRT

### ISSUE THREAD

**Issue by fiinix00** [dotnet/roslyn#42316]
**Version Used**:  Version 16.5.0 Preview 5.0 (with sdk 3.1.200-preview-015002)

**Steps to Reproduce**:

```csharp
#nullable enable

using System.Threading.Tasks;

class A
{
    static void AppUse(Func<object, Func<Task>, Task> middleware) { }

    static void B()
    {
        AppUse((context, next) =>
        {
            string? C()
            {
                return string.Concat("a", "b"); //warning
            }

            return next();
        });
    }
}
```
**Expected Behavior**:
No warning is reported

**Actual Behavior**:
warning: CS8619 Nullability of reference types in value of type 'string' doesn't match target type 'Task'.

**Comment by jcouv** [comment]
Tagging @agocke for triage
